src/twisted/protocols/test/test_tls.py

```diff
@@ -1819,3 +1833,53 @@ def acceptableProtocols(self):
         @rtype: L{list} of L{bytes}
         """
         return self._acceptableProtocols
+
+
+class AggregateSmallWritesTests(SynchronousTestCase):
+    """Tests for ``AggregateSmallWrites``."""
+
+    @given(
+        st.lists(
+            st.one_of(
+                st.none(),
+                st.integers(min_value=1, max_value=100_000).map(
+                    lambda length: (b"0123456789ABCDEFGHIJ" * ((length // 20) + 1))[
+                        :length
+                    ]
+                ),
+            ),
+            max_size=1_000,
+        )
+    )
+    def test_writes_get_aggregated(self, writes: list[Union[bytes, None]]) -> None:
```

Comment:
I am not a big fan of writing tests in this way.
It makes the test code harder to read ...and reading the test code is the only way to check the validity of the tests.

also, by writing multiple tests you can write separate tests for each scenario.

Reply:
What do you mean "this way"?

Reply:
I mean writing test code that has its logic flow, with nested if/else, without describing the use cases that are covered.

I prefer tests that just setup the system under tests for a specific scenario, the system is executed and then the result is asserted.

Also, if there is a bug for a particular use case, you should see that particular tests failing.

--------

I understand the need for parameters on tests, but I think it makes the tests harder to read.

-----

For example, is not easy to understand the scenario in which `None` is used.

Reply:
I tried refactoring it to be clearer.

Reply:
This still looks complicated to me and if there is a bug, I am not sure how I am supposed to troubleshot this test.

It's not clear how you do TDD with these style of tests. I tried to search hypothesis docs for TDD but there are no results.


It's not clear why the decorators are needed.

We have `st.none()` defined in the decorator, and the explanation is in the docstring.

To me ,it seems like the decorator can be replaced with simple function calls.

```
@given(st.integers(), st.integers())
def test_ints_are_commutative(x, y):
    assert x + y == y + x

def test_ints_are_commutative():
    x = st.integers()
    y = st.integers()
    assert x + y == y + x
```

I understand the need for things like `st.integers()` 
but I think that instead of decorators, is better to have a dedicated generator function.

I understand that for complex generators, it can lead to some boilerplace code... but this code is harder to read

Reply:
You need the decorator to ensure you run multiple versions of the inputs.

Reply:
Another way to say why there are decorators is: that's just the way Hypothesis works.

I do get @adiroiban hesitations; Hypothesis isn't always very intuitive at first (is my recollection). That said, though, it is pretty powerful and my (somewhat limited) use of it _has_ revealed surprising (to me) bugs. It's also decent at "narrowing" so it doesn't just dump out the first random failing thing (it tries to make a shorter example). I don't think Hypothesis is appropriate for every test, but it is a powerful tool and can get a lot of mileage out of a single test.

All that said, I do agree that this composite strategy is hard to read. The "none" part is hard to understand (at least a quick comment as to why -- "indicates a flush", I think? -- would help a lot).

Another way to factor this code would be to make a custom strategy -- then the docstring for that custom strategy can explain what it is generating and how/why. I would probably use `binary()` since that's at least familiar for Hypothesis users (and does something extremely similar to this code).

(Maybe I'll try to state this suggestion as a diff....)

Reply:
I pushed some updates so if you do, work off of that.